src/backend/opencl/compile_module.cpp

```diff
@@ -131,6 +152,46 @@ Module compileModule(const string &moduleKey, const vector<string> &sources,
     auto program      = opencl::buildProgram(sources, options);
     auto compileEnd   = high_resolution_clock::now();
 
+#ifdef AF_CACHE_KERNELS_TO_DISK
+    const int device             = opencl::getActiveDeviceId();
+    const string &cacheDirectory = getCacheDirectory();
+    if (!cacheDirectory.empty()) {
+        const string cacheFile = cacheDirectory + AF_PATH_SEPARATOR +
+                                 getKernelCacheFilename(device, moduleKey);
+        const string tempFile =
+            cacheDirectory + AF_PATH_SEPARATOR + makeTempFilename();
+        try {
+            // write module hash and binary data to file
+            ofstream out(tempFile, std::ios::binary);
+
+            auto binaries = program.getInfo<CL_PROGRAM_BINARIES>();
+
+            // TODO Handle cases where program objects are created from contexts
+            // having multiple devices
+            const size_t clbinSize = binaries[0].size();
+            const char *clbin =
+                reinterpret_cast<const char *>(binaries[0].data());
+            const size_t clbinHash = deterministicHash(clbin, clbinSize);
+
+            out.write(reinterpret_cast<const char *>(&clbinHash),
+                      sizeof(clbinHash));
+            out.write(reinterpret_cast<const char *>(&clbinSize),
+                      sizeof(clbinSize));
+            out.write(static_cast<const char *>(clbin), clbinSize);
+            out.close();
```

Comment:
nit: You don't need to do this because its going to be handled when out goes out of scope.

Reply:
True but the file is being renamed after this line, so we want the file write to be complete before the rename.